compiler/codegen/symbol.cpp

```diff
@@ -948,64 +962,213 @@ void TypeSymbol::codegenMetadata() {
+      im = dtImag[FLOAT_SIZE_32]->symbol;
+      fieldSize = 4;
+    } else {
+      re = dtReal[FLOAT_SIZE_64]->symbol;
+      im = dtImag[FLOAT_SIZE_64]->symbol;
+      fieldSize = 8;
+    }
+    re->codegenMetadata();
+    im->codegenMetadata();
+
+    llvm::Type *int64Ty = llvm::Type::getInt64Ty(ctx);
+    llvm::ConstantAsMetadata *zero =
+      info->mdBuilder->createConstant(llvm::ConstantInt::get(int64Ty, 0));
+    llvm::ConstantAsMetadata *fsz =
+      info->mdBuilder->createConstant(llvm::ConstantInt::get(int64Ty,
+                                                             fieldSize));
+
+    llvm::Metadata *TypeOps[5];
+    TypeOps[0] = llvm::MDString::get(ctx,cname);
+    TypeOps[1] = re->llvmTbaaTypeDescriptor;
+    TypeOps[2] = zero;  // offset
+    TypeOps[3] = im->llvmTbaaTypeDescriptor;
+    TypeOps[4] = fsz;   // offset
+    llvmTbaaAggTypeDescriptor = llvm::MDNode::get(ctx, TypeOps);
+
+    llvm::Metadata *CopyOps[6];
+    CopyOps[0] = zero;  // offset
+    CopyOps[1] = fsz;   // size
+    CopyOps[2] = re->llvmTbaaAccessTag;
+    CopyOps[3] = fsz;   // offset
+    CopyOps[4] = fsz;   // size
+    CopyOps[5] = im->llvmTbaaAccessTag;
+    llvmTbaaStructCopyNode = llvm::MDNode::get(ctx, CopyOps);
+
+    llvm::Metadata *ConstCopyOps[6];
+    ConstCopyOps[0] = zero;  // offset
+    ConstCopyOps[1] = fsz;   // size
+    ConstCopyOps[2] = re->llvmConstTbaaAccessTag;
+    ConstCopyOps[3] = fsz;   // offset
+    ConstCopyOps[4] = fsz;   // size
+    ConstCopyOps[5] = im->llvmConstTbaaAccessTag;
+    llvmConstTbaaStructCopyNode = llvm::MDNode::get(ctx, ConstCopyOps);
+
+    llvmTbaaTypeDescriptor = llvmTbaaAggTypeDescriptor;
+  } else if (!ct || hasFlag(FLAG_STAR_TUPLE) ||
+             isClass(type) || hasEitherFlag(FLAG_REF,FLAG_WIDE_REF) ||
+             hasEitherFlag(FLAG_DATA_CLASS,FLAG_WIDE_CLASS)) {
+    if (is_imag_type(type)) {
+      // At present, imaginary often aliases with real,
+      // so make real the parent of imaginary.
+      INT_ASSERT(type == dtImag[FLOAT_SIZE_32] ||
+                 type == dtImag[FLOAT_SIZE_64]);
+      TypeSymbol *re;
+      if (type == dtImag[FLOAT_SIZE_32]) {
+        re = dtReal[FLOAT_SIZE_32]->symbol;
+      } else {
+        re = dtReal[FLOAT_SIZE_64]->symbol;
+      }
+      re->codegenMetadata();
+      parent = re->llvmTbaaTypeDescriptor;
+    }
     llvmTbaaTypeDescriptor =
       info->mdBuilder->createTBAAScalarTypeNode(cname, parent);
-  } else if (ct && !isUnion(type) && !hasFlag(FLAG_STAR_TUPLE)) {
-    // Create the TBAA struct type descriptors and tbaa.struct metadata nodes.
-    llvm::SmallVector<llvm::Metadata*, 16> TypeOps;
-    llvm::SmallVector<llvm::Metadata*, 16> CopyOps;
-    llvm::SmallVector<llvm::Metadata*, 16> ConstCopyOps;
+  } else if (isRecord(type)) {
+    if (ct == dtLocaleID) {
+      // Report this as if it were a scalar type because it is a runtime type.
+      // Change this if the locale ID can contain other than 32-bit integers.
```

Comment:
It seems to me that the Chapel compiler should be able to know this? I believe you should be able to request the LLVM type information for "chpl_localeID_t" (using some of the clang integration) and then verify it only has 32-bit integers at this point. Otherwise I'm worried that a seemingly innocent change there will mess up optimization without anybody noticing.

Reply:
E.g. can you do `getTypeLLVM("chpl_localeID_t")` ? Or `codegenCType` with the result of `lvt->getType("chpl_localeID_t")` ?